File: ex5_factorial/src/Factorial.fs
open System
open System.Numerics

let factorial (n: int): BigInteger =
    let rec factorialHelper (n: BigInteger) (acc: BigInteger) =
        if n.IsZero then
            acc
        else
            factorialHelper (n - BigInteger.One) (acc * n)
    factorialHelper (BigInteger n) BigInteger.One

[<EntryPoint>]
let main argv =
    printfn "%A" (factorial 10000)
    0

File: ex5_factorial/src/factorial.kt
import java.math.BigInteger

fun factorial(n: Int): BigInteger {
    tailrec fun factorialHelper(n: BigInteger, acc: BigInteger): BigInteger {
        if (n == BigInteger.ZERO)
            return acc
        else
            return factorialHelper(n - BigInteger.ONE, acc * n)
    }

    return factorialHelper(BigInteger.valueOf(n.toLong()), BigInteger.ONE)
}

fun main(args: Array<String>) {
  print("Result = ${factorial(10000)}\n")
}
